### PR TITLE
[ticket/13817] Add core.update_session event

### PR DESCRIPTION
**Pull request by satanasov**
Add core.update_session event in \phpbb\session.php

PHPBB3-13817

### CODE REVIEW

File: phpBB/phpbb/session.php [hunk 1]
```diff
@@ -1541,12 +1541,23 @@ function unset_admin()
...
 			WHERE session_id = '" . $db->sql_escape($session_id) . "'";
 		$db->sql_query($sql);
+
+		/**
+		* Event to send session information to extension
+		*
+		* @event core.update_session
+		* @var	array		session_data				Associative array of session keys to be updated
+		* @var	array		session_id				current user's session_id
+		* @since 3.1.4
```
Comment (nickvergessen) [inline phpBB/phpbb/session.php]: `3.1.5-RC1`

File: phpBB/phpbb/session.php [hunk 2]
```diff
@@ -851,6 +851,18 @@ function session_create($user_id = false, $set_admin = false, $persist_login = f
...
 		$this->data = array_merge($this->data, $sql_ary);
 
+		/**
+		* Event to send new session data to extension
+		*
+		* @event core.session_create
+		* @var	array		session_data				Associative array of session keys to be updated
+		* @since 3.1.5-RC1
+		*/
+		$session_data = $this->data;
```
Comment (nickvergessen) [inline phpBB/phpbb/session.php]: please move before the doc block of the event

File: phpBB/phpbb/session.php [hunk 3]
```diff
@@ -906,13 +918,28 @@ function session_create($user_id = false, $set_admin = false, $persist_login = f
...
+		/**
+		* Event to send session kill information to extension
+		*
+		* @event core.session_kill
+		* @var	int		user_id				user_id of the session user.
+		* @var	string		session_id				current user's session_id
+		* @since 3.1.5-RC1
+		*/
+		$user_id = (int) $this->data['user_id'];
+		$session_id = $this->session_id;
```
Comment (nickvergessen) [inline phpBB/phpbb/session.php]: please no code between event docs and `$vars + extract()` lines

File: phpBB/phpbb/session.php [hunk 4]
```diff
@@ -906,13 +918,28 @@ function session_create($user_id = false, $set_admin = false, $persist_login = f
...
+		$vars = array('user_id', 'session_id');
```
Comment (nickvergessen) [inline phpBB/phpbb/session.php]: Please add `new_session`

File: phpBB/phpbb/session.php [hunk 5]
```diff
@@ -906,13 +918,28 @@ function session_create($user_id = false, $set_admin = false, $persist_login = f
...
 
 		$sql = 'DELETE FROM ' . SESSIONS_TABLE . "
 			WHERE session_id = '" . $db->sql_escape($this->session_id) . "'
 				AND session_user_id = " . (int) $this->data['user_id'];
 		$db->sql_query($sql);
 
+		/**
+		* Event to send session kill information to extension
+		*
+		* @event core.session_kill
```
Comment (nickvergessen) [inline phpBB/phpbb/session.php]: name should reflect that is after the session has been killed

File: phpBB/phpbb/session.php [hunk 6]
```diff
@@ -851,6 +851,18 @@ function session_create($user_id = false, $set_admin = false, $persist_login = f
 		$_SID = $this->session_id;
 		$this->data = array_merge($this->data, $sql_ary);
 
+		/**
+		* Event to send new session data to extension
+		*
+		* @event core.session_create
```
Comment (nickvergessen) [inline phpBB/phpbb/session.php]: name should reflect whether its before or after the creation of the session

File: phpBB/phpbb/session.php (hunk 2 above)
Comment (satanasov) [inline phpBB/phpbb/session.php]: Done

File: phpBB/phpbb/session.php (hunk 3 above)
Comment (satanasov) [inline phpBB/phpbb/session.php]: Done

File: phpBB/phpbb/session.php (hunk 4 above)
Comment (satanasov) [inline phpBB/phpbb/session.php]: Done

File: phpBB/phpbb/session.php (hunk 5 above)
Comment (satanasov) [inline phpBB/phpbb/session.php]: Done

File: phpBB/phpbb/session.php [hunk 7]
```diff
@@ -893,6 +893,18 @@ function session_create($user_id = false, $set_admin = false, $persist_login = f
 			$_SID = '';
 		}
 
+		$session_data = $this->data;
+		/**
+		* Event to send new session data to extension
+		*
+		* @event core.session_create_after
+		* @var	array		session_data				Associative array of session keys to be updated
```
Comment (Nicofuma) [inline phpBB/phpbb/session.php]: maybe add a note saying it is read only?

File: phpBB/phpbb/session.php [hunk 8]
```diff
@@ -906,13 +918,29 @@ function session_create($user_id = false, $set_admin = false, $persist_login = f
...
 		$db->sql_query($sql);
 
+		$user_id = (int) $this->data['user_id'];
+		$session_id = $this->session_id;
+		/**
+		* Event to send session kill information to extension
+		*
+		* @event core.session_kill_after
+		* @var	int		user_id				user_id of the session user.
+		* @var	string		session_id				current user's session_id
```
Comment (Nicofuma) [inline phpBB/phpbb/session.php]: same here, add a note to say it is read only?

File: phpBB/phpbb/session.php [hunk 9]
```diff
@@ -1541,12 +1577,23 @@ function unset_admin()
...
+		/**
+		* Event to send update session information to extension
+		*
+		* @event core.update_session
+		* @var	array		session_data				Associative array of session keys to be updated
+		* @var	string		session_id				current user's session_id
+		* @since 3.1.5-RC1
+		*/
+		$vars = array('session_data', 'session_id');
+		extract($phpbb_dispatcher->trigger_event('core.update_session', compact($vars)));
```
Comment (Nicofuma) [inline phpBB/phpbb/session.php]: Same here, read only?

Comment (satanasov) [inline phpBB/phpbb/session.php]: Done

**Pull request merged**